fl4health/losses/mkmmd_loss.py

```diff
@@ -0,0 +1,177 @@
+# multi-kernel maximum mean discrepancy
+# cao bin, HKUST, China, binjacobcao@gmail.com
+# free to charge for academic communication
+
+from typing import List, Optional
+
+import torch
+from qpth.qp import QPFunction
+
+
+class MKMMDLoss(torch.nn.Module):
+    def __init__(
+        self,
+        device: torch.device,
+        gamma_list: List[float] = [
+            2,
+            1,
+            1 / 2,
+            1 / 4,
+            1 / 8,
+        ],
+        min_distance: Optional[bool] = False,
+    ) -> None:
+        """Compute the multi-kernel maximum mean discrepancy (MK-MMD) between the source and target domains.
+        Args:
+
+            gamma_list (List[float]): list of length scales for rbf kernels
+        """
+        super().__init__()
+
+        self.kernel_num = len(gamma_list)
+        self.gamma_list = torch.tensor(gamma_list)
+        self.device = device
+        if min_distance:
+            self.sign = -1
+        else:
+            self.sign = 1
+
+    def rbf_multi_kernel(self, X1: torch.Tensor, X2: torch.Tensor) -> torch.Tensor:
+        """
+        Compute the Radial Basis Function (RBF) kernel between two sets of data points.
+
+        Args:
+            X1 (torch tensor): shape (n_samples1, n_features)
+            X2 (torch tensor): shape (n_samples2, n_features)
+            gamma (torch tensor): the kernel parameters
+
+        Returns:
+            K (torch.tensor): shape (n_samples1, n_samples2), the RBF kernel matrix
+        """
+
+        # Calculate the pairwise squared Euclidean distances
+        dist_sq = torch.sum(X1**2, dim=1, keepdim=True) + torch.sum(X2**2, dim=1) - 2 * torch.mm(X1, X2.t())
+
+        # Compute the RBF kernel matrix
+        k_list = []
+        for gamma in self.gamma_list:
+            k_list.append((torch.exp((-1 * dist_sq) / (2 * torch.pow(gamma, 2)))).unsqueeze(0))
+
+        return torch.cat(k_list)
+
+    def compute_mmd(
+        self,
+        XX: torch.Tensor,
+        YY: torch.Tensor,
+        YX: torch.Tensor,
+        XY: torch.Tensor,
+    ) -> torch.Tensor:
+
+        return XX.mean(dim=(1, 2)) + YY.mean(dim=(1, 2)) - XY.mean(dim=(1, 2)) - YX.mean(dim=(1, 2))
+
+    def estimate_η_k(
```

Comment:
I think special characters like `η` should not be used on function or variable names.